server/socketManager.js

```diff
@@ -0,0 +1,333 @@
+      this.logger.info(`dispatch "${event}" to ${id} ${clientId} client`);
+      const resp = await client.call(event, ...args);
+      return resp;
+    });
+  }
+
+  /*
+   * add clients object to map of manager's clients
+   * @param {string} id - id to store clients under
+   * @param {Object} clients
+   * @param {NodeClient} [clients.node]
+   * @param {WalletClient} [clients.wallet]
+   * @param {MultisigClient} [clients.multisig]
+   * @returns {void}
+   */
+  async addClients(id, clients) {
+    assert(typeof id === 'string', 'Must pass an id and must be a string');
+
+    if (this.clients.has(id))
+      throw new Error(`Clients with id ${id} already exists`);
+
+    for (let client in clients) {
+      await clients[client].open();
+    }
+
+    this.clients.set(id, clients);
+  }
+
+  /*
+   * Remove clients of the given id from manager
+   * @param {string} id - id of clients to remove
+   * @returns {void}
+   */
+  async removeClients(id) {
+    assert(typeof id === 'string', 'Must pass an id and must be a string');
+
+    if (!this.clients.has(id)) {
+      this.logger.debug(`No clients with id ${id} exists in manager`);
+      return;
+    }
+
+    const clients = this.clients.get(id);
+    for (let client in clients) {
+      await clients[client].close();
+    }
+
+    this.clients.delete(id);
+  }
+}
+
+class SocketManagerOptions {
+  /**
+   * SocketManagerOptions
+   * @alias module:http.SocketManagerOptions
+   * @constructor
+   * @param {Object} options
+   */
+  constructor(options) {
+    this.logger = null;
+    this.apiKey = base58.encode(random.randomBytes(20));
+    this.apiHash = sha256.digest(Buffer.from(this.apiKey, 'ascii'));
+    this.noAuth = false;
+    this.port = 8000;
+
+    this.fromOptions(options);
+  }
+
+  /**
+   * Inject properties from object.
+   * @private
+   * @param {Object} options
+   * @returns {SocketManagerOptions}
+   */
+
+  fromOptions(options) {
+    assert(options, 'Must pass options object');
+
+    if (options.logger != null) {
+      assert(typeof options.logger === 'object');
+      this.logger = options.logger;
+    }
+
+    if (options.apiKey != null) {
+      assert(typeof options.apiKey === 'string', 'API key must be a string.');
+      assert(options.apiKey.length <= 255, 'API key must be under 256 bytes.');
+      this.apiKey = options.apiKey;
+      this.apiHash = sha256.digest(Buffer.from(this.apiKey, 'ascii'));
+    }
+
+    if (options.noAuth != null) {
+      assert(typeof options.noAuth === 'boolean');
+      this.noAuth = options.noAuth;
+    }
+
+    if (options.host != null) {
+      assert(typeof options.host === 'string');
+      this.host = options.host;
+    }
+
+    if (options.port != null) {
+      assert(
+        (options.port & 0xffff) === options.port,
+        'Port must be a number.'
+      );
+      this.port = options.port;
+    }
+
+    // Allow no-auth implicitly
+    // if we're listening locally.
+    if (!options.apiKey) {
```

Comment:
I feel like there should be a log line in here when there are magic rules around changing the config without the user knowing

Reply:
this is straight JJ's :). I can add a log line though

Reply:
👍